### ISSUE TITLE
File.Copy copies through a symlink

### ISSUE THREAD

**Issue by Forgind** [dotnet/runtime#80832]
### Description

See https://github.com/dotnet/msbuild/issues/8273

It appears that when passed a symlink, File.Copy overwrites the file at the other end of the symlink rather than overwriting the symlink. Is this intentional behavior? If so, perhaps there should be an overload that just overwrites the symlink?

### Reproduction Steps

(From the issue above)
1. Build an executable project with hard or symbolic links enabled
2. Update the version of referenced NuGet package
3. Build the application again, but this time, without using hard/symbolic links.
MSBuild will try to update dependencies in the application's output folder. Sadly, instead of replacing existing links, it replaces actual files in NuGet cache (thus corrupting it).

- hard-links
```
dotnet nuget locals --clear all
dotnet new console
dotnet add package newtonsoft.json -v 13.0.1
dotnet build /p:CreateHardLinksForCopyLocalIfPossible=true
dotnet add package newtonsoft.json -v 13.0.2
dotnet build
```

- symbolic-links
```
dotnet nuget locals --clear all
dotnet new console
dotnet add package newtonsoft.json -v 13.0.1
dotnet build /p:CreateSymbolicLinksForCopyLocalIfPossible=true
dotnet add package newtonsoft.json -v 13.0.2
dotnet build
```

In both cases file `newtonsoft.json\13.0.1\lib\netstandard2.0\Newtonsoft.Json.dll` is silently replaced with `newtonsoft.json\13.0.2\lib\net6.0\Newtonsoft.Json.dll`:

### Expected behavior

Files in the NuGet cache remain untouched.

### Actual behavior

Files in a NuGet package are silently replaced with files from another version.

### Known Workarounds

Delete the file before calling Copy

**Comment by Jozkee** [comment]
Small repro: 

```cs
namespace symlink_copy;
class Program
{
    static void Main(string[] args)
    {
        File.Delete("foo");
        File.Delete("linkToFoo");

        File.WriteAllText("foo", "This is foo's content");
        File.CreateSymbolicLink("linkToFoo", "foo");

        ReadAllText("foo");
        ReadAllText("linkToFoo");

        // now, copy bar to linkToFoo. 
        File.Create("bar").Dispose();
        File.Copy("bar", "linkToFoo", overwrite: true);

        // print again
        ReadAllText("foo");
        ReadAllText("linkToFoo");
    }

    static void ReadAllText(string path)
    {
        Console.WriteLine($"{path}'s content: " + File.ReadAllText(path));
    }
}
```

**Comment by am11** [comment]
Note that this is the same behavior as `cp` command in linux without `--remove-destination` https://stackoverflow.com/q/9371222/863980 (and probably the underlying syscalls?). If it was intentional (which seems likely), then overload solution sounds better. cc @tmds

**Comment by tmds** [comment]
Yes, the behavior is intentional.

Note that with .NET 7 on Linux, File.Copy will perform copy-on-write clones on file systems that support it (like Btrfs or XFS). That matches the behavior that is desired here. (PR: https://github.com/dotnet/runtime/pull/59695)

**Comment by manfred-brands** [comment]
Does that mean the behaviour becomes filesystem dependent?

**Comment by tmds** [comment]
The functional behavior is the same. The performance is better on file systems that support the copy-on-write clone because we avoid copying the data.

**Comment by am11** [comment]
Yes, the functional behavior is the same on macOS. Output of @Jozkee's program on osx-arm64 is also:

```
foo's content: This is foo's content
linkToFoo's content: This is foo's content
foo's content: 
linkToFoo's content: 
```
